packages/backend-auth/src/reference_factory.test.ts

```diff
@@ -0,0 +1,279 @@
+      {
+        Statement: [
+          {
+            Action: 'cognito-idp:AdminCreateUser',
+            Effect: 'Allow',
+            Resource: backendAuth.resources.userPool.userPoolArn,
+          },
+        ],
+        Version: '2012-10-17',
+      }
+    );
+  });
+
+  void describe('getResourceAccessAcceptor', () => {
+    void it('attaches policies to the authenticated role', () => {
+      const backendAuth = authFactory.getInstance(getInstanceProps);
+      const testPolicy = new Policy(stack, 'testPolicy', {
+        statements: [
+          new PolicyStatement({
+            actions: ['s3:GetObject'],
+            resources: ['testBucket/testObject/*'],
+          }),
+        ],
+      });
+      const resourceAccessAcceptor = backendAuth.getResourceAccessAcceptor(
+        'authenticatedUserIamRole'
+      );
+
+      assert.equal(
+        resourceAccessAcceptor.identifier,
+        'authenticatedUserIamRoleResourceAccessAcceptor'
+      );
+
+      resourceAccessAcceptor.acceptResourceAccess(testPolicy, [
+        { name: 'test', path: 'test' },
+      ]);
+      const template = Template.fromStack(stack);
+      template.resourceCountIs('AWS::IAM::Policy', 1);
+      template.hasResourceProperties('AWS::IAM::Policy', {
+        PolicyDocument: {
+          Statement: [
+            {
+              Action: 's3:GetObject',
+              Effect: 'Allow',
+              Resource: 'testBucket/testObject/*',
+            },
+          ],
+        },
+        Roles: [backendAuth.resources.authenticatedUserIamRole.roleName],
+      });
+    });
+
+    void it('attaches policies to the unauthenticated role', () => {
+      const backendAuth = authFactory.getInstance(getInstanceProps);
+      const testPolicy = new Policy(stack, 'testPolicy', {
+        statements: [
+          new PolicyStatement({
+            actions: ['s3:GetObject'],
+            resources: ['testBucket/testObject/*'],
+          }),
+        ],
+      });
+      const resourceAccessAcceptor = backendAuth.getResourceAccessAcceptor(
+        'unauthenticatedUserIamRole'
+      );
+
+      assert.equal(
+        resourceAccessAcceptor.identifier,
+        'unauthenticatedUserIamRoleResourceAccessAcceptor'
+      );
+
+      resourceAccessAcceptor.acceptResourceAccess(testPolicy, [
+        { name: 'test', path: 'test' },
+      ]);
+      const template = Template.fromStack(stack);
+      template.resourceCountIs('AWS::IAM::Policy', 1);
+      template.hasResourceProperties('AWS::IAM::Policy', {
+        PolicyDocument: {
+          Statement: [
+            {
+              Action: 's3:GetObject',
+              Effect: 'Allow',
+              Resource: 'testBucket/testObject/*',
+            },
+          ],
+        },
+        Roles: [backendAuth.resources.unauthenticatedUserIamRole.roleName],
+      });
+    });
+  });
+});
+
+const resetFactoryCount = () => {
+  AmplifyReferenceAuthFactory.factoryCount = 0;
+};
```

Comment:
Same as in https://github.com/aws-amplify/amplify-backend/pull/2076#discussion_r1786794906

Should we have single common counter for both `defineAuth` and `referenceAuth` . I.e. make them mutually exclusive?

Reply:
updated

Reply:
thanks.

one nit: if we want to be super pedantic the counter should now live somewhere outside of both factories (it's now shared property). but feel free to keep it as is given that we have tests to guard the behavior.